### PR TITLE
Add conj function and adjoint tests

### PR DESCRIPTION
**Pull request by Randl**
Add adjoint property, which is useful for complex-valued operators.
See https://github.com/google/lineax/issues/57

### CODE REVIEW

**Comment by patrick-kidger** [comment]
What's the use-case for adding this? The only reason we have a `transpose` function is to handle the transposition necessary to synthesise the backward (VJP) pass. In particular JAX performs only transposition (in contrast to PyTorch, which performs conjugate-transposition), so I don't think we actually need this for any internal operation.

**Comment by Randl** [comment]
Specifically for normal operators, e.g., in CG part

**Comment by patrick-kidger** [comment]
Hmm. I think we can probably handle that without adding a new API. Rewriting this line:
https://github.com/google/lineax/blob/9a573529f3eea2a6944009f6e4afe4f364366baf/lineax/_solver/cg.py#L120
to be `ω(_transpose_mv(ω(_mv(vector)).call(jnp.conj).ω)).call(jnp.conj).ω` should work I think. (This is morally just a `jnp.conj(_transpose_mv(jnp.conj(_mv(vector))))`, the `ω` is just a neat syntax we have for tree-mapping an operation. And check my math but I think `conj(A conj(b)) = conj(A) b`.)

Mostly I'm trying to avoid adding more to the public API. We'd also then need to register rules for every downstream operator.

**Comment by Randl** [comment]
I see the reasoning, but currently it's suggested that the user should do the adjoint themselves at least in some cases:

https://github.com/google/lineax/blob/9a573529f3eea2a6944009f6e4afe4f364366baf/lineax/_solver/cg.py#L107-L112

Also, not having adjoint means bilinears are out of the question, if this is relevant.

**Comment by patrick-kidger** [comment]
Hmm.
Okay, I think I buy your argument. In that case I think I'd suggest adding a `lineax.conj` function, using single-dispatch in the same way as `lineax.{linearise, ...}`. This means that we'll be compatible with any other existing `AbstractLinearOperator` subclasses out there (no new abstract method), and also means we don't need to worry about the complexities of doing the full adjoint in one go -- we can decompose into conjugation and transposition, which should be simpler to reason about.

**Comment by Randl** [comment]
Ok I'll try to tackle it tomorrow.

I'm not sure the conj is easier than adjoint, specifically due to `(op @ vec.conj()).conj()` trick, which allows to find adjoint without materializing, which seems to be non-trivial for `conj`.

File: lineax/_operator.py [hunk 1]
```diff
@@ -1904,3 +1904,105 @@ def _(operator):
...
+@conj.register(TaggedLinearOperator)
+def _(operator):
+    return TaggedLinearOperator(conj(operator.operator), operator.tags)
+
+
+@conj.register(TangentLinearOperator)
+def _(operator):
+    c = lambda operator: conj(operator)
+    primal_out, tangent_out = eqx.filter_jvp(c, (operator.primal,), (operator.tangent,))
+    return TangentLinearOperator(primal_out, tangent_out)
```
Comment (patrick-kidger) [inline lineax/_operator.py:1983]: I'd need to think quite carefully whether this implementation is technically correct. I'd suggest dummying it out like so:

https://github.com/google/lineax/blob/9a573529f3eea2a6944009f6e4afe4f364366baf/lineax/_operator.py#L1759-L1765

For context, `TangentLinearOperator` is an internal detail that's only used in a few specific ways when doing JVP-of-`linear_solve`. I don't think `conj` can ever be called on it.

**Review by patrick-kidger — COMMENTED** [review]
Nits aside, this LGTM! Thank you for taking the time to make this happen.

**Comment by patrick-kidger** [comment]
By the way, let me know what your plan is regarding the various PRs. I think what might be easiest is to finish all of them before merging any of them, and then we can merge them in whatever order you think is best.

(Some good news - I have no changes planned for Lineax at the moment, so you shouldn't need to worry about keeping any of them up to date with any other changes.)

File: lineax/_operator.py [hunk 2]
```diff
@@ -1904,3 +1904,105 @@ def _(operator):
...
+    fn = _NoAuxOut(_NoAuxIn(operator.fn, operator.args))
+    jvpfn = lambda vec: jax.jvp(fn, (operator.x,), (vec,))[1]
+    jvpfnc = lambda x: jvpfn(x.conj()).conj()
+    return FunctionLinearOperator(jvpfnc, operator.in_structure(), operator.tags)
+
+
+@conj.register(FunctionLinearOperator)
+def _(operator):
+    return FunctionLinearOperator(
+        lambda vec: operator.mv(vec.conj()).conj(),
```
[the remark this answers is not in the document]
Comment (Randl) [reply lineax/_operator.py]: Can you suggest a test for that?

File: lineax/__init__.py [hunk 3]
```diff
@@ -66,6 +67,7 @@
     lower_triangular_tag as lower_triangular_tag,
     negative_semidefinite_tag as negative_semidefinite_tag,
     positive_semidefinite_tag as positive_semidefinite_tag,
+    self_adjoint_tag as self_adjoint_tag,
```
[the remark this answers is not in the document]
Comment (Randl) [reply lineax/__init__.py:70]: Hmm, it was there to optimize out `adjoint`, but after splitting into conj+transpose that's not possible. Should we give it up?

**Comment by Randl** [comment]
From my perspective, these 5 PRs are ready to merge (up to minor unclosed discussions). It does not fully implement complex support, but it would be simpler for me to continue when these are merged so I can be sure this part already works.

As for the merge order, I think it should #59 and #63 should go first, then #60, then #61 and this one

File: lineax/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Randl) [reply lineax/__init__.py:70]: Or maybe we can add `adjoint` function which checks the tag before doing `conj(op.T)`

File: lineax/_operator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply lineax/_operator.py]: Something like:
```python
def fn(y, args):
    return {"b": y["a"]}

y_struct = jax.eval_shape(lambda: {"a": 0.0})
operator = FunctionLinearOperator(fn, y_struct)
```
which has a pytree-valued input (`{"a": *}`) and a pytree-valued output (`{"b": *}`).

File: lineax/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply lineax/__init__.py:70]: I think I'd probably just give it up.

**Comment by patrick-kidger** [comment]
Also LGTM!

**Pull request merged**